### PR TITLE
[Pipelines] Support passing pipeline param to function spec resources

### PR DESCRIPTION
**Pull request by Tankilevitch**
https://jira.iguazeng.com/browse/ML-3713

Until now we weren't able to pass pipeline param to function spec due to two reasons:
- strict regex validation that was applied on function attributes specifically resources attribute
- when running a function inside a pipeline we were storing the function to the DB and holding only the function url inside the workflow spec, this caused failure in converting the pipeline param to actual value because that this transformation happens when kfp transforms the kfp step to actual k8s resource, therefor we had to add logic that if we find pipelineparam in the function spec we don't save the function to the db but rather saves the full function spec in the workflow spec so that pipeline will convert that placeholder to value.

Actions made -
- Added an option to manage more lose validation on regex which allows to have "any" ( at least one ) match on regex
- Added logic that looks for `{{pipelineparam:op` in function spec while in workflow generation and if exists saves the whole function spec in the workflow spec, so that kfp will transform that placeholder to a value.

### CODE REVIEW

**Review by theSaarco — CHANGES_REQUESTED** [review]
Looks good, with some minor comments.

File: tests/integration/sdk_api/httpdb/test_exception_handling.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def test_exception_handling(self):
             match=rf"400 Client Error: Bad Request for url: http:\/\/(.*)\/{mlrun.get_run_db().get_api_path_prefix()}"
             r"\/projects: Failed creating project some_p"
             r"roject details: MLRunInvalidArgumentError\(\"Field \'project\.metadata\.name\' is malformed"
-            r"\. Does not match required pattern: (.*)\"\)",
+            r"\. some_project does not match required pattern: (.*)\"\)",
```
Comment (alonmr) [inline tests/integration/sdk_api/httpdb/test_exception_handling.py:58]:
```suggestion
            rf"\. {invalid_project_name} does not match required pattern: (.*)\"\)",
```

File: mlrun/runtimes/base.py [hunk 2]
```diff
@@ -599,6 +599,9 @@ def _force_handler(self, handler):
         if not handler:
             raise RunError(f"handler must be provided for {self.kind} runtime")
 
+    def _has_pipeline_param(self) -> bool:
+        return self.to_json().find("{{pipelineparam:op") > 0
```
Comment (alonmr) [inline mlrun/runtimes/base.py:603]: please add a comment that points to pipeline_param regex for more info/full regex (I was mainly confused by why the brackets are not closed)

File: tests/system/projects/test_project.py [hunk 3]
```diff
@@ -485,6 +485,25 @@ def test_local_pipeline(self):
...
+        project.set_function(
+            name="func-1",
+            func=code_path,
+            kind="job",
+            image="mlrun/mlrun",
+            handler="handler",
+        )
+        # set and run a two-step workflow in the project
+        project.set_workflow("paramflow", workflow_path)
+        project.run("paramflow", engine="kfp", watch=True)
```
Comment (alonmr) [inline tests/system/projects/test_project.py:505]: I thought you would be passing a memory param and verify that it was passed correctly.
Did I misunderstand the feature?

**Review by alonmr — COMMENTED** [review]
minor comments/questions

File: mlrun/runtimes/base.py [hunk 4]
```diff
@@ -678,10 +681,9 @@ def as_step(
         :return: KubeFlow containerOp
         """
 
-        # if self.spec.image and not image:
-        #     image = self.full_image_path()
-
-        if use_db:
+        # if the function spec contain KFP PipelineParams (futures) pass the full spec to the
+        # ContainerOp this way KFP will substitute the params with previous step outputs
+        if (use_db or use_db is None) and not self._has_pipeline_param():
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/runtimes/base.py:686]: fixed

File: mlrun/runtimes/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/runtimes/base.py:603]: i'll update the comment, as I don't want to strict the user

Comment (Tankilevitch) [reply mlrun/runtimes/base.py:603]: re-used the regex pipelineparam and added reference to it in the regex file

File: tests/system/projects/test_project.py (hunk 3 above)
Comment (Tankilevitch) [reply tests/system/projects/test_project.py:505]: it had default value which the pipeline used, but took your comment and added explicit parameter

**Review by theSaarco — APPROVED** [review]
Looks good 👍

**Review by liranbg — APPROVED** [review]

**Review by alonmr — APPROVED** [review]
🔥

**Pull request merged**